### PR TITLE
docs: update datasets tutorial

### PR DESCRIPTION
**Pull request by SauravMaheshkar**
This PR aims to update and refactor Tutorial 1 "Structure your Input"

### CODE REVIEW

File: docs/source/tutorials/structure_your_input.rst [hunk 1]
```diff
@@ -161,11 +109,175 @@ An example for an input directory with videos could look like this:
...
+        assert "image" in example_batch.keys(), "batch should be of type Dict[str, Any] with a key 'image'"
+
+        example_batch["image"] = [
+            transform(image.convert("RGB")) for image in example_batch["image"]
+        ]
+        return example_batch
+
+    base.set_transform(apply_transform)
+
+    hf_dataloader = torch.utils.data.DataLoader(base["train"])
```
Comment (MalteEbner) [inline docs/source/tutorials/structure_your_input.rst:183]: I would make the function shorter.
```suggestion
    import torch
    from typing import Any
    from datasets import load_dataset
    from lightly.transforms import SimCLRTransform

    base = load_dataset("uoft-cs/cifar10", trust_remote_code=True)

    ## Use pre-defined set of transformations from Lightly SSL
    transform = SimCLRTransform()

    def apply_transform(batch: dict[str, Any], transform=transform)-> Dict[str, Any]:
        """
        Applies the given transform on all elements of batch["image"].
        """
        assert "image" in example_batch, "batch must contain key 'image'"
        batch["image"] = [transform(img.convert("RGB")) for img in batch["image"]]
        return batch

    base.set_transform(apply_transform)

    hf_dataloader = torch.utils.data.DataLoader(base["train"])
```

**Review by MalteEbner — COMMENTED** [review]
Looks good, I like it!

What I am most unsure about: Did you try actually feeding a `hf_dataloader` into the Lightly SSL training scripts? They might be incompatible, as the Lightly SSL training scripts expect a (batch, targets, filenames) dataloader output, which is not provided by huggingface.

File: docs/source/tutorials/structure_your_input.rst [hunk 2]
```diff
@@ -3,93 +3,63 @@
...
-jump right into training a model:
-
-- :ref:`lightly-moco-tutorial-2`
-- :ref:`lightly-simclr-tutorial-3`  
-- :ref:`lightly-simsiam-tutorial-4`
-- :ref:`lightly-custom-augmentation-5`
-- :ref:`lightly-detectron-tutorial-6`
+The modern-day open-source ecosystem has changed a lot over the years, and there are now
+many viable options for data pipelining. For experimentation purposes,
+the `torchvision.data <https://pytorch.org/vision/main/datasets.html>`_ submodule provides a robust implementation for most use cases,
```
[the remark this answers is not in the document]
Comment (SauravMaheshkar) [reply docs/source/tutorials/structure_your_input.rst:8]: It seems like there was some miscommunication here. I didn't mean to convey that torchvision is "experimental", but rather that torchvision hosts a lot of smaller datasets which can be loaded on-device for verifying training pipelines on smaller datasets.

File: docs/source/tutorials/structure_your_input.rst [hunk 3]
```diff
@@ -3,93 +3,63 @@
...
+and the `HuggingFace Hub <https://hf.co>`_ has emerged as a growing collection of datasets that span a variety of domains and tasks.
+When you have your own data in memory, the ability to quickly create datasets and dataloaders is of prime importance.
```
[the remark this answers is not in the document]
Comment (SauravMaheshkar) [reply docs/source/tutorials/structure_your_input.rst:10]: Thanks, makes sense.

File: docs/source/tutorials/structure_your_input.rst (hunk 1 above)
Comment (SauravMaheshkar) [reply docs/source/tutorials/structure_your_input.rst:183]: Thank you, that is much cleaner 🧼

**Comment by SauravMaheshkar** [comment]
> g scripts? They might be incompatible, as the Lightly SSL training scripts expect a (batch, targets, filenames) dataloader output, which is not provided by huggingface.

Yes, I've tried it out and it does work with Lightly Training Scripts. The Colab Notebook released with the recent vector index blogpost uses the food101 dataset from the Huggingface for training.

**Review by MalteEbner — APPROVED** [review]
Looks good from my side :)

File: docs/source/tutorials/structure_your_input.rst [hunk 4]
```diff
@@ -3,93 +3,61 @@
...
-- :ref:`lightly-simsiam-tutorial-4`
-- :ref:`lightly-custom-augmentation-5`
-- :ref:`lightly-detectron-tutorial-6`
-
-If you are looking for a use case that's not covered by the above tutorials please
-let us know by `creating an issue <https://github.com/lightly-ai/lightly/issues/new>`_
-for it.
+In this blog post, we will provide a brief overview of the `LightlyDataset <https://docs.lightly.ai/self-supervised-learning/lightly.data.html#lightly.data.dataset.LightlyDataset>`_
+and go through examples of using datasets from various open-source libraries such as PyTorch and
+Huggingface with the Lightly SSL open-source package. We will also look into how we can create dataloaders
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:13]:
```suggestion
Huggingface with Lightly SSL. We will also look into how we can create dataloaders
```

File: docs/source/tutorials/structure_your_input.rst [hunk 5]
```diff
@@ -3,93 +3,61 @@
...
+for video tasks while incorporating weak labels.
 
 
-Supported File Types
---------------------
+Native LightlyDataset format
+----------------------------
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:18]:
```suggestion
LightlyDataset
--------------
```

File: docs/source/tutorials/structure_your_input.rst [hunk 6]
```diff
@@ -3,93 +3,61 @@
...
 
-By default, the `Lightly SSL Python package <https://pypi.org/project/lightly/>`_ 
-can process images or videos for self-supervised learning or for generating embeddings.
-You can always write your own torch-like dataset to use other file types.
+The LightlyDataset class aims to provide a uniform data interface for all models and functions in the Lightly SSL package.
+It allows us to create both Image and Video dataset classes with or without labels.
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:21]:
```suggestion
It allows us to create both image and video dataset classes with or without labels.
```

File: docs/source/tutorials/structure_your_input.rst [hunk 7]
```diff
@@ -3,93 +3,61 @@
...
 - .mov, .mp4 and .avi
 
+Unlabeled Image Datasets
+^^^^^^^^^^^^^^^^^^^^^^^^^
 
-Image Folder Datasets
----------------------
-
-Image folder datasets contain raw images and are typically specified with the `input_dir` key-word.
+Creating an Unlabeled Image Dataset is the simplest and the most common use case.
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:41]:
```suggestion
Creating an unlabeled image dataset is the simplest and the most common use case.
```

File: docs/source/tutorials/structure_your_input.rst [hunk 8]
```diff
@@ -3,93 +3,61 @@
...
 
+Assuming all images are in a single directory, you can simply pass in the path to the directory containing all the images, viz.
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:43]:
```suggestion
Assuming all images are in a single directory, you can simply pass in the path to the directory containing all the images.
```

File: docs/source/tutorials/structure_your_input.rst [hunk 9]
```diff
@@ -3,93 +3,61 @@
...
-- :ref:`lightly-moco-tutorial-2`
-- :ref:`lightly-simclr-tutorial-3`  
-- :ref:`lightly-simsiam-tutorial-4`
-- :ref:`lightly-custom-augmentation-5`
-- :ref:`lightly-detectron-tutorial-6`
-
-If you are looking for a use case that's not covered by the above tutorials please
-let us know by `creating an issue <https://github.com/lightly-ai/lightly/issues/new>`_
-for it.
+In this blog post, we will provide a brief overview of the `LightlyDataset <https://docs.lightly.ai/self-supervised-learning/lightly.data.html#lightly.data.dataset.LightlyDataset>`_
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:11]:
```suggestion
In this tutorial, we will provide a brief overview of the `LightlyDataset <https://docs.lightly.ai/self-supervised-learning/lightly.data.html#lightly.data.dataset.LightlyDataset>`_
```

File: docs/source/tutorials/structure_your_input.rst [hunk 10]
```diff
@@ -3,93 +3,61 @@
...
   `many more <https://pillow.readthedocs.io/en/stable/handbook/image-file-formats.html>`_
 
-Videos
-^^^^^^^^^^^^^^^^^^^^^
-
-To load videos directly, Lightly SSL uses 
-`torchvision <https://github.com/pytorch/vision>`_ and 
+To load videos directly, Lightly SSL uses
+`torchvision <https://github.com/pytorch/vision>`_ and
 `PyAV <https://github.com/PyAV-Org/PyAV>`_. The following formats are supported.
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:34]:
```suggestion
Lightly SSL uses
`torchvision <https://github.com/pytorch/vision>`_ and
`PyAV <https://github.com/PyAV-Org/PyAV>`_ for video loading. The following formats are supported.
```

File: docs/source/tutorials/structure_your_input.rst [hunk 11]
```diff
@@ -3,93 +3,61 @@
...
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Labeled Image Datasets
+^^^^^^^^^^^^^^^^^^^^^^
 
-You can give structure to your input directory by collecting the input images in subdirectories. In this case,
-the filenames loaded by Lightly SSL are with respect to the "root directory" `data/`. Furthermore, Lightly SSL assigns
-each image a so-called "weak-label" indicating to which subdirectory it belongs.
+If you have (weak) labels for your images and if each label has its directory,
+then you can simply pass in the path to the parent directory,
+and the dataset class will assign each image its subdirectory as label., viz.
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:60]:
```suggestion
and the dataset class will assign each image its subdirectory as label.
```

File: docs/source/tutorials/structure_your_input.rst [hunk 12]
```diff
@@ -114,41 +82,19 @@ each image a so-called "weak-label" indicating to which subdirectory it belongs.
...
-        0,
-        ...
-        0,
-        1,
-        ...
-        1,
-        ...
-        9,
-    ]
+   labeled_dataset = LightlyDataset(input_dir='labeled_images_dir/')
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:89]:
```suggestion
   labeled_dataset = LightlyDataset(input_dir='labeled_image_dir/')
```

Please also rename the `data/` dir in the example directory structure to `labeled_image_dir`

File: docs/source/tutorials/structure_your_input.rst [hunk 13]
```diff
@@ -114,41 +82,19 @@ each image a so-called "weak-label" indicating to which subdirectory it belongs.
...
----------------------
-The Lightly SSL Python package allows you to work `directly` on video data, without having
-to exctract the frames first. This can save a lot of disk space as video files are
-typically strongly compressed. Using Lightly SSL on video data is as simple as pointing 
-the software at an input directory where one or more videos are stored. The package will
-automatically detect all video files and index them so that each frame can be accessed.
+The Lightly SSL package also has native support for videos (`.mov`, `.mp4`, and `.avi` file extensions are supported),
+without having to extract the frames first. This can save a lot of disk space as video files are
+typically strongly compressed. No matter if your videos are in one flat directory or distributed across subdirectories,
+you can simply pass the path into the LightlyDataset constructor.
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:97]: Please rename `data/` in example directory structure to `video_dir/`

File: docs/source/tutorials/structure_your_input.rst [hunk 14]
```diff
@@ -161,11 +107,187 @@ An example for an input directory with videos could look like this:
...
+
+   from lightly.data import LightlyDataset
+
+   video_dataset = LightlyDataset(input_dir='video_dir/')
+
+The dataset assigns each video frame its video as label.
+
+.. note::
+
+   To use video-specific features of Lightly SSL download the necessary extra dependencies `pip install lightly"[video]"`. Furthermore,
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:120]:
```suggestion
   To use video-specific features of Lightly SSL download the necessary extra dependencies `pip install "lightly[video]"`. Furthermore,
```

File: docs/source/tutorials/structure_your_input.rst [hunk 15]
```diff
@@ -161,11 +107,187 @@ An example for an input directory with videos could look like this:
...
+   randomly accessing video frames is slower compared to accessing the extracted frames on disk. However,
+   by working directly on video files, one can save a lot of disk space because the frames do not have to
+   be extracted beforehand.
+
+PyTorch Datasets
+----------------
+
+You can also use native `torchvision <https://pytorch.org/vision/main/datasets.html>`_ datasets with Lightly SSL directly.
+Just create a dataset as you normally would and apply transforms for greater control over the dataloader. For example, BYOL
+(Bootstrap your own element) expects two views of an input image. To achieve this, we can use the `BYOLTransform`
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:130]:
```suggestion
Just create a dataset as you normally would and apply transforms for greater control over the dataloader. For example, the BYOL self-supervised learning method expects two views of an input image. To achieve this, we can use the `BYOLTransform`
```

You could also make BYOL link to the BYOL docs.

File: docs/source/tutorials/structure_your_input.rst [hunk 16]
```diff
@@ -161,11 +107,187 @@ An example for an input directory with videos could look like this:
...
+while creating the dataset instance, which will lead to the dataloader returning two views per batch, viz.
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:131]:
```suggestion
while creating the dataset instance, which will lead to the dataloader returning two views per batch.
```

File: docs/source/tutorials/structure_your_input.rst [hunk 17]
```diff
@@ -161,11 +107,187 @@ An example for an input directory with videos could look like this:
...
+
+.. code-block:: python
+
+   import torchvision
+   from lightly.transforms.byol_transform import (
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:136]:
```suggestion
   from lightly.transforms import (
```

File: docs/source/tutorials/structure_your_input.rst [hunk 18]
```diff
@@ -161,11 +107,187 @@ An example for an input directory with videos could look like this:
...
+       BYOLTransform,
+       BYOLView1Transform,
+       BYOLView2Transform,
+   )
+
+   transform = BYOLTransform(
+       view_1_transform=BYOLView1Transform(input_size=32, gaussian_blur=0.0),
+       view_2_transform=BYOLView2Transform(input_size=32, gaussian_blur=0.0),
+   )
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:145]: BYOL is maybe not a great example as it requires more boilerplate code for creating the transformer compared to SimCLR. Consider switching to SimCLR

File: docs/source/tutorials/structure_your_input.rst [hunk 19]
```diff
@@ -161,11 +107,187 @@ An example for an input directory with videos could look like this:
...
+       batch_size=256,
+       shuffle=True,
+       drop_last=True,
+       num_workers=8,
+   )
+
+   # ...
+   for batch in dataloader:
+       x0, x1 = batch[0]
+       # ...
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:161]: I don't think this part is necessary, we also don't show it for LightlyDataset

File: docs/source/tutorials/structure_your_input.rst [hunk 20]
```diff
@@ -161,11 +107,187 @@ An example for an input directory with videos could look like this:
...
+
+
+HuggingFace Datasets
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:164]:
```suggestion
Hugging Face Datasets
```

Please also correct in other places on this page

File: docs/source/tutorials/structure_your_input.rst [hunk 21]
```diff
@@ -161,11 +107,187 @@ An example for an input directory with videos could look like this:
...
+--------------------
+
+To use a dataset from the huggingface hub 🤗, we can simply apply the desired transformations using the
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:167]:
```suggestion
To use a dataset from the Hugging Face Hub 🤗, we can simply apply the desired transformations using the 
```

File: docs/source/tutorials/structure_your_input.rst [hunk 22]
```diff
@@ -161,11 +107,187 @@ An example for an input directory with videos could look like this:
...
+
+
+.. code-block:: python
+
+    import torch
+    from typing import Any
+    from datasets import load_dataset
+    from lightly.transforms import SimCLRTransform
+
+    base = load_dataset("uoft-cs/cifar10", trust_remote_code=True)
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:179]:
```suggestion
    dataset = load_dataset("uoft-cs/cifar10", trust_remote_code=True)
```

File: docs/source/tutorials/structure_your_input.rst [hunk 23]
```diff
@@ -161,11 +107,187 @@ An example for an input directory with videos could look like this:
...
+
+    ## Use pre-defined set of transformations from Lightly SSL
+    transform = SimCLRTransform()
+
+    def apply_transform(batch: dict[str, Any], transform=transform)-> Dict[str, Any]:
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:184]:
```suggestion
    def apply_transform(batch: Dict[str, Any])-> Dict[str, Any]:
```
Also make sure to import `Dict` from typing

File: docs/source/tutorials/structure_your_input.rst [hunk 24]
```diff
@@ -161,11 +107,187 @@ An example for an input directory with videos could look like this:
...
+        """
+        Applies the given transform on all elements of batch["image"].
+        """
+        assert "image" in example_batch, "batch must contain key 'image'"
+        batch["image"] = [transform(img.convert("RGB")) for img in batch["image"]]
+        return batch
+
+    base.set_transform(apply_transform)
+    hf_dataloader = torch.utils.data.DataLoader(base["train"])
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:193]:
```suggestion
    dataset.set_transform(apply_transform)
    dataloader = torch.utils.data.DataLoader(dataset["train"])
```

File: docs/source/tutorials/structure_your_input.rst [hunk 25]
```diff
@@ -161,11 +107,187 @@ An example for an input directory with videos could look like this:
...
+-------------------
+
+Many SSL methods leverage image augmentations to better learn invariances in the training process. For example,
+by using different crops of a given image, the SSL model will be trained to produce a representation
+that is invariant to these different crops. When using a operation such as grayscale or colorjitter as augmentation,
+it will produce a representation that is invariant to the color information [1]_.
+
+We can use off the shelf augmentations from libraries like `torchvision transforms <https://pytorch.org/vision/stable/transforms.html>`_
+and `albumentations <https://albumentations.ai/docs/>`_ or the ones offered by Lightly SSL's
+`transforms <https://docs.lightly.ai/self-supervised-learning/lightly.transforms.html>`_ submodule while creating our datasets, viz.
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:205]:
```suggestion
We can use off the shelf augmentations from libraries like `torchvision transforms <https://pytorch.org/vision/stable/transforms.html>`_
and `albumentations <https://albumentations.ai/docs/>`_ or the ones offered by Lightly SSL's
`transforms <https://docs.lightly.ai/self-supervised-learning/lightly.transforms.html>`_ submodule while creating our datasets.
```

File: docs/source/tutorials/structure_your_input.rst [hunk 26]
```diff
@@ -3,93 +3,61 @@
...
-    +--- img-2.jpg
-    ...
-    +--- img-N.jpg
+.. code-block:: python
 
-For the structure above, Lightly SSL will understand the input as follows:
+    from lightly.data import LightlyDataset
 
-.. code-block:: python
+    dataset = LightlyDataset(input_dir='image_dir/')
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:49]: I would also include a transform here as we show transforms also for pytorch and hugging face.

File: docs/source/tutorials/structure_your_input.rst [hunk 27]
```diff
@@ -161,11 +107,187 @@ An example for an input directory with videos could look like this:
...
+
+    ## Lightly Transforms
+    lightly_transform = SimCLRTransform()
 
 .. note::
 
-    Randomly accessing video frames is slower compared to accessing the extracted frames on disk. However,
-    by working directly on video files, one can save a lot of disk space because the frames do not have to
-    be extracted beforehand.
+   You can also create your own augmentations, for more details please refer to :ref:`lightly-custom-augmentation-5`
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:235]:
```suggestion
   You can also create your own SSL augmentations, for more details please refer to :ref:`lightly-custom-augmentation-5` 
```

File: docs/source/tutorials/structure_your_input.rst [hunk 28]
```diff
@@ -161,11 +107,187 @@ An example for an input directory with videos could look like this:
...
+    ## Torchvision Transforms
+    torchvision_transforms = T.Compose(
+        [
+            T.RandomHorizontalFlip(),
+            T.ToTensor(),
+        ]
+    )
+
+    ## Albumentation Transforms
+    albumentation_transforms = A.Compose(
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:222]:
```suggestion
    albumentation_transform = A.Compose(
```

File: docs/source/tutorials/structure_your_input.rst [hunk 29]
```diff
@@ -161,11 +107,187 @@ An example for an input directory with videos could look like this:
...
+`transforms <https://docs.lightly.ai/self-supervised-learning/lightly.transforms.html>`_ submodule while creating our datasets, viz.
+
+.. code-block:: python
+
+    import albumentations as A
+    import torchvision.transforms as T
+    from albumentations.pytorch import ToTensorV2
+
+    ## Torchvision Transforms
+    torchvision_transforms = T.Compose(
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:214]:
```suggestion
    torchvision_transform = T.Compose(
```

File: docs/source/tutorials/structure_your_input.rst [hunk 30]
```diff
@@ -161,11 +107,187 @@ An example for an input directory with videos could look like this:
...
+
+    torchvision_aug_dataset = LightlyDataset.from_torch_dataset(
+        base, transform=torchvision_transforms
+    )
+    albumentation_aug_dataset = LightlyDataset.from_torch_dataset(
+        base, transform=albumentation_transforms
+    )
+    lightly_aug_dataset = LightlyDataset.from_torch_dataset(
+        base, transform=lightly_transform
+    )
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:270]: I think these two sections are a bit overkill as they are quite verbose but do not add much new information. I would replace them and just mention that we can use the transforms for any dataset as shown previously and give a minimal example like:
```
dataset = LightlyDataset(input_dir="image_dir/", transform=torchvision_transform)
# or
dataset = torchvision.datasets.CIFAR10("datasets/cifar10", download=True, transform=albumentations_transform)
```

File: docs/source/tutorials/structure_your_input.rst [hunk 31]
```diff
@@ -161,11 +107,187 @@ An example for an input directory with videos could look like this:
...
+
+Conclusion
+----------
+
+In this blogpost, we went through examples of using various open-source packages to create datasets and dataloaders with Lightly SSL,
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:275]:
```suggestion
In this tutorial, we went through examples of using various open-source packages to create datasets and dataloaders with Lightly SSL,
```

**Review by guarin — CHANGES_REQUESTED** [review]
Thanks for adding all this helpful information! I left a bunch of comments, mostly regarding structure and formatting.

File: docs/source/tutorials/structure_your_input.rst (hunk 26 above)
Comment (guarin) [reply docs/source/tutorials/structure_your_input.rst:48]: This hasn't been addressed yet

File: docs/source/tutorials/structure_your_input.rst [hunk 32]
```diff
@@ -114,58 +81,162 @@ each image a so-called "weak-label" indicating to which subdirectory it belongs.
...
+            A.CenterCrop(height=128, width=128),
+            A.Normalize(mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225)),
+            ToTensorV2(),
+        ]
+    )
+
+    ## Lightly Transforms
+    lightly_transform = SimCLRTransform()
+
+    dataset = LightlyDataset(input_dir="image_dir/", transform=torchvision_transform)
```
Comment (guarin) [inline docs/source/tutorials/structure_your_input.rst:214]: Suggestion to make it more evident that datasets/transforms are interchangeable:

```suggestion
    
    # Datasets and transforms can be mixed and matched together:
    dataset = LightlyDataset(input_dir="image_dir/", transform=torchvision_transform)
    dataset = torchvision.datasets.CIFAR10("datasets/cifar10", transform=lightly_transform)
```

**Review by guarin — COMMENTED** [review]
Looks great! Two small things left and then we can merge :)

**Review by guarin — APPROVED** [review]
LGTM! Thanks a lot!

**Pull request merged**